pkg/abstract.go

```diff
@@ -83,8 +84,15 @@ func scrapeAwsData(config ScrapeConf, now time.Time, metricsPerQuery int, fips,
 			}
 		}
 	}
+
+	go func() {
```

Comment:
I think there's no guarantee here that this will iterate on all the errors and append them before the function returns right? And similarly this function doesn't know when there will not be any more errors coming through (so it can safely block)

Reply:
If I understand the intent here (to dump the errors from the channel to the slice), could we potentially move this to after `wg.Wait()`?

Reply:
>  think there's no guarantee here that this will iterate on all the errors and append them before the function returns right? And similarly this function doesn't know when there will not be any more errors coming through (so it can safely block)

I can't quite tell if you're saying there's an issue or saying it's okay. My thought was since I'm using an unbuffered channel [docs](https://golang.org/doc/effective_go#channels) "If the channel is unbuffered, the sender blocks until the receiver has received the value." So, I should be able to guarantee that this iterates on all the errors before the function returns because we wait for the waitgroup to finish and then close the err channel which should cause the range to exit. Is there a better way to accomplish this?

> If I understand the intent here (to dump the errors from the channel to the slice), could we potentially move this to after wg.Wait()?

If I understand channels correctly, in order to do that I would have to have a channel with a large enough buffer to account for all expected errors to prevent send from blocking. This might be Jobs * Roles * Regions but it doesn't feel like a safe assumption to make.

Reply:
The go routine isn't guaranteed to be completed by the time the function returns though even if we move `wg.Wait()` to after the function invocation, as it's not blocking.

From what I'm understanding, the worker routines will block, and therefore not call done, until they can write the error to the channel, but the channel will be empty before the error is written to the slice - so there is potential that we won't get all the errors.

As an example, hopefully it is clear:
3 workers (w1, w2, w3)
**w1**:
- sends error
- done

**w2**:
- blocks sending error 

errorRoutine:
- reads error (from w1)
- appends to slice

**w2**:
- now can send error (channel free)
- done

**w3**:
- blocks sending error  

errorRoutine:
- reads error (from w2)

**w3**:
- now can send error (channel free)
- done

**errorRoutine**:
- appends to slice

**function returns** -> All workers done

- reads error (from w3)

From what I can tell, we need to guarantee (through some coordination) that the errorRoutine has finished processing all the errors in some way, otherwise it is _possible_ that it won't be finished

Reply:
Sorry @kgeckhart I didn't refresh my page to see your reply - hopefully my message above ^ still makes sense.

>  So, I should be able to guarantee that this iterates on all the errors before the function returns because we wait for the waitgroup to finish which would mean all errors would need to be flushed. After the waitgroup finishes closing the err channel should cause the range to exit. Is there a better way to accomplish this?

We guarantee all errors have been written to the channel, but not that the errors have been written to the slice as may return before that go routine has exited

Reply:
You are indeed correct that we can miss errors here. I reworked this to show in goplayground, https://play.golang.org/p/inktbXhTs1v. You're right I'll need another WaitGroup or a done channel to fix this.

Reply:
Nice example 👍 Yep that makes sense to add another in this case

Reply:
Okay, I added a WaitGroup for errors so now it's structured like https://play.golang.org/p/xOo9wzORmv_l which appears to be good. How does that look?